lib/stdlib/src/erl_tar.hrl

```diff
@@ -0,0 +1,383 @@
+          real_size :: binary()}). % integer
+-type header_gnu() :: #header_gnu{}.
+%% The set of fields specific to STAR-formatted archives
+-record(header_star, {
+          header_v7 :: header_v7(),
+          magic :: binary(),
+          version :: binary(),
+          uname :: binary(),
+          gname :: binary(),
+          devmajor :: binary(), % integer
+          devminor :: binary(), % integer
+          prefix :: binary(),
+          atime :: binary(), % integer
+          ctime :: binary(), % integer
+          trailer :: binary()}).
+-type header_star() :: #header_star{}.
+%% The set of fields specific to USTAR-formatted archives
+-record(header_ustar, {
+          header_v7 :: header_v7(),
+          magic :: binary(),
+          version :: binary(),
+          uname :: binary(),
+          gname :: binary(),
+          devmajor :: binary(), % integer
+          devminor :: binary(), % integer
+          prefix :: binary()}).
+-type header_ustar() :: #header_ustar{}.
+
+-type header_fields() :: header_v7() |
+                         header_gnu() |
+                         header_star() |
+                         header_ustar().
+
+%% The overall tar reader, it holds the low-level file handle,
+%% it's access, position, and the I/O primitives wrapper.
+-record(reader, {
+          handle :: file:io_device() | term(),
+          access :: read | write | ram,
+          pos = 0 :: non_neg_integer(),
+          func :: file_op()
+         }).
+-type reader() :: #reader{}.
+%% A reader for a regular file within the tar archive,
+%% It tracks it's current state relative to that file.
+-record(reg_file_reader, {
+          handle :: reader(),
+          num_bytes = 0,
+          pos = 0,
+          size = 0
+         }).
+-type reg_file_reader() :: #reg_file_reader{}.
+%% A reader for a sparse file within the tar archive,
+%% It tracks it's current state relative to that file.
+-record(sparse_file_reader, {
+          handle :: reader(),
+          num_bytes = 0, % bytes remaining
+          pos = 0, % pos
+          size = 0, % total size of file
+          sparse_map = #sparse_array{}
+         }).
+-type sparse_file_reader() :: #sparse_file_reader{}.
+
+%% Types for the readers
+-type reader_type() :: reader() | reg_file_reader() | sparse_file_reader().
+-type handle() :: file:io_device() | term().
+
+%% Types for the I/O primitive wrapper functions
+-type write_fun() :: fun((write, {handle(), iodata()}) ->
+                                ok | {error, term()}).
+-type close_fun() :: fun((close, handle()) ->
+                                ok | {error, term()}).
+-type read_fun() :: fun((read2, {handle(), non_neg_integer()}) ->
+                               {ok, string() | binary()} | eof | {error, term()}).
+-type position_fun() :: fun((position, {file:io_device() | term, non_neg_integer()}) ->
+                                   {ok, non_neg_integer()} | {error, term()}).
+-type file_op() :: write_fun()
```

Comment:
Ah interesting, ok, I'll think on this one a bit and update it. I'm not sure if specifying `file_op()` as you've described is useful as a typespec, since it's unlikely to catch the mistakes I was attempting to catch.